libs/libarchfpga/src/read_xml_arch_file.cpp

```diff
@@ -5041,3 +5169,986 @@ static T* get_type_by_name(const char* type_name, std::vector<T>& types) {
+        if ((router_info->second.first == 1) && (router_info->second.second == 0)) {
+            archfpga_throw("", -1,
+                           "The router with id:'%d' is not connected to any other router in the NoC.", router_info->first);
+
+        } // case where a router was found to be connected to another router but not declared using the <router> tag in the arch file (ie. missing)
+        else if ((router_info->second.first == 0) && (router_info->second.second > 0)) {
+            archfpga_throw("", -1,
+                           "The router with id:'%d' was found to be connected to another router but missing in the architecture file. Add the router using the <router> tag.", router_info->first);
+
+        } // case where the router was delcared multiple times in the architecture file (multiple <router> tags for the same router)
+        else if (router_info->second.first > 1) {
+            archfpga_throw("", -1,
+                           "The router with id:'%d' was included more than once in the architecture file. Routers should only be declared once.", router_info->first);
+        }
+    }
+
+    return;
+}
+
+/* for vib arch*/
+static void ProcessVibArch(pugi::xml_node Parent, std::vector<t_physical_tile_type>& PhysicalTileTypes, t_arch* arch, const pugiutil::loc_data& loc_data) {
+    pugi::xml_node Node;
+    //pugi::xml_node SubElem;
+
+    //arch->is_vib_arch = true;
+    int num_vibs = count_children(Parent, "vib", loc_data);
+    arch->vib_infs.reserve(num_vibs);
+    Node = get_first_child(Parent, "vib", loc_data);
+    
+    for (int i_vib = 0; i_vib < num_vibs; i_vib++) {
+        ProcessVib(Node, PhysicalTileTypes, arch, loc_data);
+        Node = Node.next_sibling(Node.name());
+    }
+}
+
+static void ProcessVib(pugi::xml_node Vib_node, std::vector<t_physical_tile_type>& PhysicalTileTypes, t_arch* arch, const pugiutil::loc_data& loc_data) {
+    pugi::xml_node Node;
+    pugi::xml_node SubElem;
+    const char* tmp;
+    int itmp;
+
+    VibInf vib;
+    std::vector<t_segment_inf> segments = arch->Segments;
+    t_arch_switch_inf* switches = arch->Switches;
+
+    tmp = get_attribute(Vib_node, "name", loc_data).as_string(nullptr);
+    if (tmp) {
+        vib.set_name(tmp);
+    }
+    else {
+        archfpga_throw(loc_data.filename_c_str(), loc_data.line(Vib_node),
+                       "No name specified for the vib!\n");
+    }
+
+    tmp = get_attribute(Vib_node, "pbtype_name", loc_data).as_string(nullptr);
+    if (tmp) {
+        vib.set_pbtype_name(tmp);
+    }
+    else {
+        archfpga_throw(loc_data.filename_c_str(), loc_data.line(Vib_node),
+                       "No pbtype_name specified for the vib!\n");
+    }
+
+    vib.set_seg_group_num(get_attribute(Vib_node, "vib_seg_group", loc_data).as_int(1));
+
+    tmp = get_attribute(Vib_node, "arch_vib_switch", loc_data).as_string(nullptr);
+    
+    if (tmp) {
+        for (int i_switch = 0; i_switch < arch->num_switches; i_switch++) {
```

Comment:
This requires the switch information is loaded (parsed) already. It could be risky once the parsing sequence is changed. Data linking should be started after all the XML is loaded.